pandas/core/reshape/merge.py

```diff
@@ -2388,6 +2390,13 @@ def _factorize_keys(
             # "_values_for_factorize"
             rk, _ = rk._values_for_factorize()  # type: ignore[union-attr]
 
+    if needs_i8_conversion(lk.dtype) and lk.dtype == rk.dtype:
+        # GH#23917 TODO: Needs tests for non-matching dtypes
+        # GH#23917 TODO: needs tests for case where lk is integer-dtype
+        #  and rk is datetime-dtype
+        lk = np.asarray(lk, dtype=np.int64)
+        rk = np.asarray(rk, dtype=np.int64)
```

Comment:
This part was (accidentally?) removed in https://github.com/pandas-dev/pandas/pull/49876. So there won't be any tests failing from leaving this out (because falling back to the object factorizer also gives correct results), but it is necessary to ensure we use the faster factorizer for datetimelikes

Reply:
i expect that the relevant cases should go through the `elif isinstance(lk, ExtensionArray) and lk.dtype == rk.dtype:` branch above?

Reply:
Not fully:

1) I did go through there before but only for tz-naive data (and same unit), now we already convert to a numpy array in the `if isinstance(lk.dtype, DatetimeTZDtype) ...` block by getting the `_ndarray` of lk and rk  (but we could not do it there, and let it go through the block that calls `_values_for_factorize`)
2) even when going through this block, the ``DatetimeArray._values_for_factorize`` returns the `_ndarray`, which is the M8 numpy array, not an integer numpy array. So we still need this `needs_i8_conversion`. 
  Now, we should maybe change `_values_for_factorize`? (but that's probably more for something targetting 2.1, and didn't check the other places where `_values_for_factorize` is being used)

Reply:
makes sense, thanks.